File: mesmaeker-core/src/main/java/fr/xgouchet/mesmaeker/core/ValidatingUtils.java
package fr.xgouchet.mesmaeker.core;

import java.util.function.Consumer;
import java.util.function.Function;

/**
 * @author Xavier F. Gouchet
 */
public class ValidatingUtils {

    public static <T, O> void verifyFunctionThrows(T subject, Function<T, O> function, Throwable expected) {
        boolean match = false;
        Throwable thrown = null;
        try {
            function.apply(subject);
        } catch (Throwable t) {
            thrown = t;
            match = expected.getClass().isAssignableFrom(t.getClass());
        }

        if (thrown == null) {
            throw new AssertionError("Expecting " + expected.getClass().getSimpleName() + " to be thrown");
        } else if (!match) {
            throw new AssertionError("Expecting " + expected.getClass().getSimpleName() + " to be thrown, but was " + thrown.getClass().getSimpleName());
        }
    }

    public static <T, O> void verifyFunctionThrows(T subject, Function<T, O> function, Class<? extends Throwable> expected) {
        boolean match = false;
        Throwable thrown = null;
        try {
            function.apply(subject);
        } catch (Throwable t) {
            thrown = t;
            match = expected.isAssignableFrom(t.getClass());
        }

        if (thrown == null) {
            throw new AssertionError("Expecting " + expected.getSimpleName() + " to be thrown");
        } else if (!match) {
            throw new AssertionError("Expecting " + expected.getSimpleName() + " to be thrown, but was " + thrown.getClass().getSimpleName());
        }
    }

    public static <T, O> void verifyConsumerThrows(T subject, Consumer<T> consumer, Class<? extends Throwable> expected) {
        boolean match = false;
        Throwable thrown = null;
        try {
            consumer.accept(subject);
        } catch (Throwable t) {
            thrown = t;
            match = expected.isAssignableFrom(t.getClass());
        }

        if (thrown == null) {
            throw new AssertionError("Expecting " + expected.getSimpleName() + " to be thrown");
        } else if (!match) {
            throw new AssertionError("Expecting " + expected.getSimpleName() + " to be thrown, but was " + thrown.getClass().getSimpleName());
        }
    }
}

File: mesmaeker-core/src/main/java/fr/xgouchet/mesmaeker/core/ValidatingUtils.kt
package fr.xgouchet.mesmaeker.core


/**
 * @author Xavier F. Gouchet
 */
object ValidatingUtils {

    @JvmStatic
    fun <T, O> verifyFunctionThrows(subject: T,
                                    function: (T) -> O,
                                    expected: Throwable) {
        var match = false
        var thrown: Throwable? = null
        try {
            function(subject)
        } catch (t: Throwable) {
            thrown = t
            match = expected.javaClass.isAssignableFrom(t.javaClass)
        }

        if (thrown == null) {
            throw AssertionError("Expecting " + expected.javaClass.simpleName + " to be thrown")
        } else if (!match) {
            throw AssertionError("Expecting " + expected.javaClass.simpleName + " to be thrown, but was " + thrown.javaClass.simpleName)
        }
    }

    @JvmStatic
    fun <T, O> verifyFunctionThrows(subject: T,
                                    function: (T) -> O,
                                    expected: Class<out Throwable>) {
        var match = false
        var thrown: Throwable? = null
        try {
            function(subject)
        } catch (t: Throwable) {
            thrown = t
            match = expected.isAssignableFrom(t.javaClass)
        }

        if (thrown == null) {
            throw AssertionError("Expecting " + expected.simpleName + " to be thrown")
        } else if (!match) {
            throw AssertionError("Expecting " + expected.simpleName + " to be thrown, but was " + thrown.javaClass.simpleName)
        }
    }

    @JvmStatic
    fun <T> verifyBlockThrows(subject: T, block: (T) -> Unit, expected: Class<out Throwable>) {
        var match = false
        var thrown: Throwable? = null
        try {
            block(subject)
        } catch (t: Throwable) {
            thrown = t
            match = expected.isAssignableFrom(t.javaClass)
        }

        if (thrown == null) {
            throw AssertionError("Expecting " + expected.simpleName + " to be thrown")
        } else if (!match) {
            throw AssertionError("Expecting " + expected.simpleName + " to be thrown, but was " + thrown.javaClass.simpleName)
        }
    }
}
